tsdb/docs/format/head_chunks.md

```diff
@@ -0,0 +1,43 @@
+# Head Chunks on Disk Format
+
+The following describes the format of a chunks file,
+which is created in the `wal/chunks/` inside the data directory.
+
+Chunks in the files are referenced from the index by uint64 composed of
+in-file offset (lower 4 bytes) and segment sequence number (upper 4 bytes).
+
+The mint and maxt in the file header is the maximum and minimum times
+among all the chunks that it holds.
+
+```
+┌──────────────────────────────┐
+│  magic(0x85BD40DD) <4 byte>  │
+├──────────────────────────────┤
+│    version(1) <1 byte>       │
+├──────────────────────────────┤
+│    padding(0) <3 byte>       │
+├──────────────────────────────┤
+│    mint <8 byte, uint64>     │
+├──────────────────────────────┤
+│    maxt <8 byte, uint64>     │
+├──────────────────────────────┤
+│ ┌──────────────────────────┐ │
+│ │         Chunk 1          │ │
+│ ├──────────────────────────┤ │
+│ │          ...             │ │
+│ ├──────────────────────────┤ │
+│ │         Chunk N          │ │
+│ └──────────────────────────┘ │
+└──────────────────────────────┘
+```
+
+
+# Chunk
+
+Unlike chunks in the on-disk blocks, here we additionally store series reference that the chunks belongs to and the mint/maxt of the chunks. This is because we don't have an index associated with these chunks, hence these meta information are used while replaying the chunks.
```

Comment:
Yes, mint maxt is obtained from chunks itself. And right, we would be replaying all the chunks on start, so we should be able to calculate mint-maxt from there (or maybe just maxt is enough after that). I will add it in the comments stating that `IterateAllChunks` should be called after initializing the struct to set those properly (there is where the iteration happens)

Reply:
I think I misunderstood your comment. If you meant getting mint and maxt of chunks from the data itself, it would not scale. We would need to iterate each and every sample to do that. When we want to restore the chunks in the memory, we need to know the mint and maxt of the chunks so that we can perform queries over them, hence we need them here.
PS: Even for persistent blocks, we store the mint and maxt separately in the index.

If you meant the above, then with this misunderstanding I actually got rid of mint and maxt for the file. As we would iterate the chunks once anyway on startup for the replay, I am restoring  the mint maxt of the files during that time. This would mean calling `IterateChunks` at least once (which we will do).

Reply:
yep that was my point, thanks , that simplifies the implementation a bit.

Reply:
> If you meant the above, then with this misunderstanding I actually got rid of mint and maxt for the file. As we would iterate the chunks once anyway on startup for the replay, I am restoring the mint maxt of the files during that time. This would mean calling IterateChunks at least once (which we will do).

I am confused, this comments reads that you have removed mint and maxt from the file, but it is still there in the code?

I am missunderstanding something

Reply:
https://github.com/prometheus/prometheus/blob/012161d90d6a8a6bb930b90601fb89ff6cc3ae60/tsdb/head.go#L2017-L2020

So a chunk does not have mint and maxt with it explicitly, it's just in the samples. So, to fill in mint and maxt in `memChunk` or the future mmaped chunk, we need this mint and maxt. 

Similarly for blocks, the mint and maxt of chunks are stored for all chunks in the index https://github.com/prometheus/prometheus/blob/012161d90d6a8a6bb930b90601fb89ff6cc3ae60/tsdb/docs/format/index.md#series

Reply:
so I guess that is used for quering mainly? So you know whih chunks to include when requesting some data?

Reply:
Yes. (And in case of mmaping, also to know max time of the file for truncation)